net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,186 @@
+	// as provided by the DERPMap.
+	DERPMapPreferred EndpointProvider = iota
+	// Tailscale is used for endpoints that are the Tailscale coordination server or admin console.
+	Tailscale
+	// Platform is used for an endpoint that is a well-known captive portal detection URL for the current platform
+	// (operated by Apple, Microsoft, etc.)
+	Platform
+	// DERPMapOther is used for an endpoint that is a DERP node, but not contained in the current preferred DERP region.
+	DERPMapOther
+)
+
+func (p EndpointProvider) String() string {
+	switch p {
+	case DERPMapPreferred:
+		return "DERPMapPreferred"
+	case Tailscale:
+		return "Tailscale"
+	case Platform:
+		return "Platform"
+	case DERPMapOther:
+		return "DERPMapOther"
+	default:
+		return fmt.Sprintf("EndpointProvider(%d)", p)
+	}
+}
+
+// Endpoint represents a URL that can be used to detect a captive portal, along with the expected
+// result of the HTTP request.
+type Endpoint struct {
+	// URL is the URL that we make an HTTP request to as part of the captive portal detection process.
+	URL string
+	// StatusCode is the expected HTTP status code that we expect to see in the response.
+	StatusCode int
+	// ExpectedContent is a string that we expect to see contained in the response body. If this is non-empty,
+	// we will check that the response body contains this string. If it is empty, we will not check the response body
+	// and only check the status code.
+	ExpectedContent string
+	// SupportsTailscaleChallenge is true if the endpoint will return the sent value of the X-Tailscale-Challenge
+	// HTTP header in its HTTP response.
+	SupportsTailscaleChallenge bool
+	// Provider is the source of the endpoint. This is used to prioritize certain endpoints over others
+	// (for example, a DERP node in the preferred region should always be used first).
+	Provider EndpointProvider
+}
+
+func (e Endpoint) String() string {
+	return fmt.Sprintf("Endpoint{URL=%q, StatusCode=%d, ExpectedContent=%q, SupportsTailscaleChallenge=%v, Provider=%s}", e.URL, e.StatusCode, e.ExpectedContent, e.SupportsTailscaleChallenge, e.Provider.String())
+}
+
+// availableEndpoints returns a set of Endpoints which can be used for captive portal detection by performing
+// one or more HTTP requests and looking at the response. The returned Endpoints are ordered by preference,
+// with the most preferred Endpoint being the first in the slice.
+func availableEndpoints(derpMap *tailcfg.DERPMap, preferredDERPRegionID int, logf logger.Logf, goos string) []Endpoint {
+	if logf == nil {
+		logf = log.Printf
+	}
+
+	endpoints := []Endpoint{}
+
+	if derpMap != nil {
+		// If we have a DERP map, we can use the DERP hostnames and IPs as captive portal detection endpoints.
+		for _, region := range derpMap.Regions {
+			for _, node := range region.Nodes {
+				if node.IPv4 == "" {
+					continue
+				}
+				url := "http://" + node.IPv4 + "/generate_204"
```

Comment:
Typically, we use `node.HostName` instead of the bare IP... but I assume that this is to ensure that we send a request over IPv4 for checking such captive portals?

Reply:
Correct, we cannot rely on DNS resolution being available as we don't know if the captive portal is intercepting DNS queries by rewriting them, or dropping them entirely.

Reply:
IPv4 isn't always populated. It's an optional hint.